flocker/node/agents/gce.py

```diff
@@ -0,0 +1,385 @@
+            blockdevice_id.
+        - Set the description of the disk resource to:
+            "flocker-v1-cluster: <cluster-uuid>".
+        - Whenever attaching a disk to an instance, set <token> to the
+            blockdevice_id.
+
+    Design implications:
+        - The GCE disk names meet the length and format requirements, and are
+            as unique as blockdevice_ids.
+        - We can perform operations on a disk only knowing its blockdevice_id
+            without making additional API calls or looking things up in a
+            table.
+        - dataset_id is a pure function of blockdevice_id and vice versa.
+        - You can have multiple clusters within the same project.
+        - Multiple clusters within the same project cannot have datasets with
+            the same UUID.
+        - We could add filtering by cluster by filtering on description.
+        - The path of the device (or at least the path to a symlink to a path
+            of the volume) is a pure function of blockdevice_id.
+    """
+    # TODO(mewert): Logging throughout.
+
+    def __init__(self, cluster_id, project, zone):
+        """
+        Initialize the GCEBlockDeviceAPI.
+
+        :param unicode project: The project where all GCE operations will take
+            place.
+        :param unicode zone: The zone where all GCE operations will take place.
+        """
+        # TODO(mewert): Also enable credentials via service account private
+        # keys.
+        credentials = AppAssertionCredentials(
+            "https://www.googleapis.com/auth/cloud-platform")
+        self._compute = discovery.build(
+            'compute', 'v1', credentials=credentials)
+        self._project = project
+        self._zone = zone
+        self._cluster_id = cluster_id
+
+    def _disk_resource_description(self):
+        """
+        Returns the value to be used in the description field of the disk
+        resources for this cluster.
+
+        :returns unicode: The value for the description.
+        """
+        return u"flocker-v1-cluster-id: " + unicode(self._cluster_id)
+
+    def _do_blocking_operation(self, function, **kwargs):
+        """
+        Perform a GCE operation, blocking until the operation completes.
+
+        This will call `function` with the passed in keyword arguments plus
+        additional keyword arguments for project and zone which come from the
+        private member variables with the same name. It is expected that
+        `function` returns an object that has an `execute()` method that
+        returns a GCE operation resource dict.
+
+        This function will then poll the operation until it reaches state
+        'DONE' or times out, and then returns the final operation resource
+        dict.
+
+        :param function: Callable that takes keyword arguments project and
+            zone, and returns an executable that results in a GCE operation
+            resource dict as described above.
```

Comment:
:question: Is it worth defining a local interface for `IExecutable`? I don't think so.